### DISCUSSION TITLE
New experiment says Hello World despite completely different code

### DISCUSSION THREAD

**Discussion by tmintz** [l3atbc-datadog/pushkin#115] [General] [answered]
I followed the "Simple Experiment" tutorial for a local test, but instead of the lexical decision experiment I used a simple SRT experiment I threw together for the jsPsych workshop.  I created a new experiment using `pushkin install experiment` and called it `srt`. I copied the correct portion of the jsPsych experiment into the new experiment.js file that was created, getting rid of all the Hello World code that was there by default. I then modified the relevant index.js file to load in the jspsych-serial-reaction-time-mouse plugin.  I ran `pushkin prep` and `pushkin start;`.  I navigated to the local site, and the srt experiment showed up on the home page, but when I clicked it, it ran the Hello World script, like all the other dummy experiments.  I then stopped pushkin and ran `pushkin armageddon` and reran prep and start, and the same thing happened.  Note, I *never* ran `pushkin prep` when the Hello World code was in the srt experiment experiment.js file -- I copied in the new code right after the file was created, so I am really at a loss.

Any ideas?

**Comment by jkhartshorne** [comment, marked as answer]
Huh. Weird. Let me try to reproduce.

**Reply by parkerrobbins** [reply]
@tmintz, I don't have your code, but, I did try reproducing this using the [SRT tutorial](https://www.jspsych.org/tutorials/rt-task/) on jspsych.org. After I edited the files and ran `pushkin prep` and `pushkin start;`, the experiment ran as expected. Would you like to post the relevant sections of your `experiment.js` and `index.js` files for us to look at?

**Reply by tmintz** [reply]
Sure.  First the experiment.js (please don't ask about the keyword names :-), then the index.js.  Note that I tried two different ways of adding the srt plugin using pushkin.loadScripts (two calls vs. two entries in the array argument) -- both yielded the same behavior.  Also, I did try adding the `lex` experiment following the tutorial, and that worked fine.

```
import jsPsych from 'pushkin-jspsych';

const timeline = []


var targets = [
    {achoo: [0,jsPsych.randomization.sampleWithoutReplacement([0,1,2,3],1)[0]],data: {seq:1}},
    {achoo: [0,jsPsych.randomization.sampleWithoutReplacement([0,1,2,3],1)[0]],data: {seq:2}},
    {achoo: [0,jsPsych.randomization.sampleWithoutReplacement([0,1,2,3],1)[0]],data: {seq:3}},
    {achoo: [0,jsPsych.randomization.sampleWithoutReplacement([0,1,2,3],1)[0]],data: {seq:4}},
    {achoo: [0,jsPsych.randomization.sampleWithoutReplacement([0,1,2,3],1)[0]],data: {seq:5}},
    {achoo: [0,jsPsych.randomization.sampleWithoutReplacement([0,1,2,3],1)[0]],data: {seq:6}},
    {achoo: [0,jsPsych.randomization.sampleWithoutReplacement([0,1,2,3],1)[0]],data: {seq:8}}
]

var trials = {
    timeline: [{
        type: 'serial-reaction-time-mouse',
        grid: [[1, 1, 1, 1]],
        target: jsPsych.timelineVariable('achoo'),
        data: jsPsych.timelineVariable('data')
    }],
    timeline_variables: targets,
    randomize_order: false,
    repetitions: 6
}

timeline.push(trials);

export default timeline;
```
and
```
import React from 'react';
import pushkinClient from 'pushkin-client';
import jsPsych from 'pushkin-jspsych';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import timeline_basic from './experiment';
import jsYaml from 'js-yaml';
const fs = require('fs');

//stylin'
import './assets/experiment.css'

const expConfig = jsYaml.safeLoad(fs.readFileSync('../config.yaml'), 'utf8');

const pushkin = new pushkinClient();
window.jsPsych = jsPsych;


const mapStateToProps = state => {
  return {
    userID: state.userInfo.userID
  };
}

class quizComponent extends React.Component {

  constructor(props) {
    super(props);
    this.state = { loading: true };
  }

  componentDidMount() {
    this.startExperiment();
  }

  async startExperiment() {
    this.setState({ experimentStarted: true });

    jsPsych.data.addProperties({user_id: this.props.userID}); //See https://www.jspsych.org/core_library/jspsych-data/#jspsychdataaddproperties
    await pushkin.connect('/api/srt');
    await pushkin.prepExperimentRun(this.props.userID);
    await pushkin.loadScripts([
      'https://cdn.jsdelivr.net/gh/jspsych/jsPsych@6.0.4/plugins/jspsych-html-keyboard-response.js',
      'https://cdn.jsdelivr.net/gh/jspsych/jsPsych@6.0.4/plugins/jspsych-serial-reaction-time-mouse.js'
    ]);
   /*  await pushkin.loadScripts([
      'https://cdn.jsdelivr.net/gh/jspsych/jsPsych@6.0.4/plugins/jspsych-serial-reaction-time-mouse.js',
     ]);
      */
    const timeline = pushkin.setSaveAfterEachStimulus(timeline_basic);
    await jsPsych.init({
      display_element: document.getElementById('jsPsychTarget'),
      timeline: timeline,
      on_finish: this.endExperiment.bind(this),
    });

    document.getElementById('jsPsychTarget').focus();
    this.setState({ loading: false });
  }

  async endExperiment() {
    document.getElementById("jsPsychTarget").innerHTML = "Processing...";
    await pushkin.tabulateAndPostResults(this.props.userID, expConfig.experimentName)
    document.getElementById("jsPsychTarget").innerHTML = "Thank you for participating!";
  }

  render() {
    const { match } = this.props;

    return (
      <div>
        {this.state.loading && <h1>Loading...</h1>}
        <div id="jsPsychTarget" />
      </div>
    );
  }
}
```

**Reply by parkerrobbins** [reply]
@tmintz In `index.js`, `export default withRouter(connect(mapStateToProps)(quizComponent));` should be the last line but was somehow deleted. Once I added it back in and re-ran `pushkin prep` and `pushkin start;`, it worked as expected. Please let me know if that solves the issue for you.

**Reply by tmintz** [reply]
Sorry, I mistakenly didn't select it in the cut and paste, but it is in index.js file. So, I'm afraid that's not the issue for me.  Is there any debug info I can give you (from Docker??) that might help?  I'm puzzled why it should be running the Hello World code, even if something was broken.  I tried running in incognito mode as well, in case there was some weird cache issue, but the behavior was the same.

**Reply by parkerrobbins** [reply]
That's probably more @jkhartshorne's area. Is there anything useful in the console when you open the developer tools?

**Reply by parkerrobbins** [reply]
Also, @tmintz your operating system (complete with version) and the output of the following may be helpful:
- `pushkin --version`
- `docker --version` 
- `docker-compose --version`

**Reply by tmintz** [reply]
RE: javascript console, there are some warnings, but they appear to be no different than the ones that are generated by `lex` from the tutorial.
```
$ pushkin --version 
1.1.4

$ docker --version
Docker version 19.03.12, build 48a66213fe

$ docker-compose --version
docker-compose version 1.26.2, build eefe0d31
```
I might just try starting from scratch.

**Reply by parkerrobbins** [reply]
Try upgrading the Pushkin CLI using the following command:
```bash
$ yarn global upgrade pushkin-cli
```
When you're done, you should expect this:
```bash
$ pushkin --version
1.1.5
```

**Reply by tmintz** [reply]
I upgraded the CLI, ran `kill`, `prep`, and `start`,  and had the same problem.  However, I created a new basic experiment, went through the same procedure for changing it to my srt experiment, and *this one* worked!  But, running `diff` on the working and non-working experiments' `experiment.js` and `index.js` reveals no differences.  I suppose it must have been something about the way the old CLI set things up in some other part of the experiment structure?  Anyway, at least I was able to achieve what I wanted.  Thanks for your help.  If you are interested in looking more into this I'd be happy to provide whatever further information I can, but my immediate problem appears to be solved.

**Reply by parkerrobbins** [reply]
Thanks for your reply! Glad you've got it working now!

**Comment by chiara-gambi** [comment]
Hi there, I am having the same issue as described in this thread - except it happens with the Simple Experiment Tutorial from the doc pages. I have tried updating to the latest version of the CLI (2.0.0), pushkin kill, pushkin armageddon and then pushkin prep but I am still only seeing the "hello world" message instead of the lexical decision experiment when I click on the lex experiment container. Anything else I can try? Thanks!

**Reply by parkerrobbins** [reply]
Hi, @chiara-gambi. Could you show me your `experiment.js` file?

**Reply by chiara-gambi** [reply]
Hi @parkerrobbins - thanks for your reply. Here is my experiment.js file

`import jsPsych from 'pushkin-jspsych';

const timeline = []

var welcome = {
    type: 'html-keyboard-response',
    stimulus: '<p>Welcome to the experiment. Please press C to continue.</p>',
    choices: ['c']
}

timeline.push(welcome);

var instructions = {
    type: 'html-keyboard-response',
    stimulus: '<p>You will see two sets of letters displayed in a box, like this:</p>'+
      '<div class="fixation"><p class="top">HELLO</p><p class="bottom">WORLD</p></div>'+
      '<p>Press Y if both sets are valid English words. Press N if one or both is not a word.</p>'+
      '<p>Press Y to continue.</p>',
    choices: ['y']
}

timeline.push(instructions);

var instructions_2 = {
    type: 'html-keyboard-response',
    stimulus: '<p>In this case you would press N</p>'+
      '<div class="fixation"><p class="top">FOOB</p><p class="bottom">ARTIST</p></div>'+
      '<p>Press N to continue to the start of the experiment.</p>',
    choices: ['n']
}

timeline.push(instructions_2);

var trials = [
    {word_1: 'SOCKS', word_2: 'SHOE', both_words: true, related: true},
    {word_1: 'SLOW', word_2: 'FAST', both_words: true, related: true},
    {word_1: 'QUEEN', word_2: 'KING', both_words: true, related: true},
    {word_1: 'LEAF', word_2: 'TREE', both_words: true, related: true},

    {word_1: 'SOCKS', word_2: 'TREE', both_words: true, related: false},
    {word_1: 'SLOW', word_2: 'SHOE', both_words: true, related: false},
    {word_1: 'QUEEN', word_2: 'FAST', both_words: true, related: false},
    {word_1: 'LEAF', word_2: 'KING', both_words: true, related: false},

    {word_1: 'AGAIN', word_2: 'PLAW', both_words: false, related: false},
    {word_1: 'BOARD', word_2: 'TRUDE', both_words: false, related: false},
    {word_1: 'LIBE', word_2: 'HAIR', both_words: false, related: false},
    {word_1: 'MOCKET', word_2: 'MEET', both_words: false, related: false},
    
    {word_1: 'FLAFF', word_2: 'PLAW', both_words: false, related: false},
    {word_1: 'BALT', word_2: 'TRUDE', both_words: false, related: false},
    {word_1: 'LIBE', word_2: 'NUNE', both_words: false, related: false},
    {word_1: 'MOCKET', word_2: 'FULLOW', both_words: false, related: false}
]

var lexical_decision_procedure = {
    timeline: [
      {
        type: 'html-keyboard-response',
        stimulus: '<div class="fixation"></div>',
        choices: jsPsych.NO_KEYS,
        trial_duration: 1000
      },
      {
        type: 'html-keyboard-response',
        stimulus: function(){
          return '<div class="fixation"><p class="top">'+jsPsych.timelineVariable('word_1', true)+'</p><p class="bottom">'+jsPsych.timelineVariable('word_2', true)+'</p></div>';
        },
        choices: ['y','n'],
        data: {
          both_words: jsPsych.timelineVariable('both_words'),
          related: jsPsych.timelineVariable('related')
        },
        on_finish: function(data){
          var char_resp = jsPsych.pluginAPI.convertKeyCodeToKeyCharacter(data.key_press);
          if(data.both_words){
            data.correct = char_resp == 'y';
          } else {
            data.correct = char_resp == 'n';
          }
        }
      },
      {
        type: 'html-keyboard-response',
        stimulus: function(){
          var last_correct = jsPsych.data.get().last(1).values()[0].correct;
          if(last_correct){
            return '<div class="fixation correct"></div>';
          } else {
            return '<div class="fixation incorrect"></div>';
          }
        },
        choices: jsPsych.NO_KEYS,
        trial_duration: 2000
      }
    ],
    timeline_variables: trials,
    randomize_order: true
}

timeline.push(lexical_decision_procedure);

var data_summary = {
    type: 'html-keyboard-response',
    stimulus: function(){
      var mean_rt_related = jsPsych.data.get().filter({related:true, both_words:true, correct: true}).select('rt').mean();
      var mean_rt_unrelated = jsPsych.data.get().filter({related:false, both_words:true, correct: true}).select('rt').mean();
      return '<p>Average response time for related words: '+Math.round(mean_rt_related)+'ms</p>'+
        '<p>Average response time for unrelated words: '+Math.round(mean_rt_unrelated)+'ms</p>'
    },
    choices: jsPsych.NO_KEYS
}

timeline.push(data_summary);

export default timeline;`

But it turns out that I missed out a crucial step in @tmintz solution. When I deleted the lex folder from my local machine and created a new experiment from scratch (after upgrading the CLI) then it worked! Before I was just running pushkin kill etc on the old lex directory built under the old version of the CLI. So that seemed to make the difference. Sorry for not realising earlier.

**Reply by parkerrobbins** [reply]
No worries! Glad it's working now!
